pkg/e2/v1beta1/resolver.go

```diff
@@ -0,0 +1,165 @@
+// SPDX-FileCopyrightText: 2020-present Open Networking Foundation <info@opennetworking.org>
+//
+// SPDX-License-Identifier: LicenseRef-ONF-Member-1.0
+
+package e2
+
+import (
+	"context"
+	"fmt"
+	"github.com/onosproject/onos-api/go/onos/topo"
+	"github.com/onosproject/onos-lib-go/pkg/grpc/retry"
+	"github.com/onosproject/onos-lib-go/pkg/uri"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/attributes"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+const resolverName = "e2"
+
+func newResolver(nodeID NodeID) resolver.Builder {
+	return &ResolverBuilder{
+		nodeID: nodeID,
+	}
+}
+
+type ResolverBuilder struct {
+	nodeID NodeID
+}
+
+func (b *ResolverBuilder) Scheme() string {
+	return resolverName
+}
+
+func (b *ResolverBuilder) Build(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+	var dialOpts []grpc.DialOption
+	if opts.DialCreds != nil {
+		dialOpts = append(
+			dialOpts,
+			grpc.WithTransportCredentials(opts.DialCreds),
+		)
+	} else {
+		dialOpts = append(dialOpts, grpc.WithInsecure())
+	}
+	dialOpts = append(dialOpts, grpc.WithUnaryInterceptor(retry.RetryingUnaryClientInterceptor(retry.WithRetryOn(codes.Unavailable, codes.Unknown))))
+	dialOpts = append(dialOpts, grpc.WithStreamInterceptor(retry.RetryingStreamClientInterceptor(retry.WithRetryOn(codes.Unavailable, codes.Unknown))))
+	dialOpts = append(dialOpts, grpc.WithContextDialer(opts.Dialer))
+
+	resolverConn, err := grpc.Dial(target.Endpoint, dialOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceConfig := cc.ParseServiceConfig(
+		fmt.Sprintf(`{"loadBalancingConfig":[{"%s":{}}]}`, resolverName),
+	)
+
+	resolver := &Resolver{
+		nodeID:        b.nodeID,
+		clientConn:    cc,
+		resolverConn:  resolverConn,
+		serviceConfig: serviceConfig,
+	}
+	err = resolver.start()
+	if err != nil {
+		return nil, err
+	}
+	return resolver, nil
+}
+
+var _ resolver.Builder = (*ResolverBuilder)(nil)
+
+type Resolver struct {
+	nodeID        NodeID
+	clientConn    resolver.ClientConn
+	resolverConn  *grpc.ClientConn
+	serviceConfig *serviceconfig.ParseResult
+}
+
+func (r *Resolver) start() error {
+	client := topo.NewTopoClient(r.resolverConn)
+	request := &topo.WatchRequest{}
+	stream, err := client.Watch(context.Background(), request)
+	if err != nil {
+		return err
+	}
+	go func() {
+		nodeURI := uri.NewURI(uri.WithScheme("e2"), uri.WithOpaque(string(r.nodeID)))
```

Comment:
Need to double check whether this URI is being constructed correctly, or if the node ID is already in the form of a URI.